omnipaxos/src/ballot_leader_election.rs

```diff
@@ -138,25 +201,16 @@ impl BallotLeaderElection {
     }*/
 
     fn check_leader(&mut self) -> Option<Ballot> {
-        self.quorum_connected = true;
         let ballots = std::mem::take(&mut self.ballots);
         let top_ballot = ballots
             .into_iter()
-            .filter_map(
-                |(ballot, candidate)| {
-                    if candidate {
-                        Some(ballot)
-                    } else {
-                        None
-                    }
-                },
-            )
+            .filter(|ballot| self.quorum.is_accept_quorum(ballot.connectivity as usize))
             .max()
             .unwrap_or_default();
 
         if top_ballot < self.leader.unwrap_or_default() {
```

Comment:
Do we not run into problems with the priority then? For example
```rust
// New incoming top_ballot
let top_ballot = BLEBallot {
    n: 1,
    priority: 0,
    connectivity: 5,
    pid: 2,
};
// The full BLEBallot of the leader when it was elected
let leader = BLEBallot {
    n: 1,
    priority: 5,
    connectivity: 5,
    pid: 1,
}
```
The `top_ballot` shouldn't become leader but it will be considered greater because we ignore the priority fields (Due to to storing `self.leader` only as a `Ballot`)?

Reply:
After thinking more about this, I think it will be hard to try and ignore the priority field. Because unlike with connectivity where we do a filter, the priority is used to order ballots. And if we would ignore the priority sometimes, it would mean ignoring a higher ballot, which is incorrect (the QLE properties state that we eventually should agree on the highest ballot).

I would like to hear what you think, but I believe the best is to revert back to having `priority` as a field in `Ballot` and have `leader: Ballot` in BLE. This will mean that if a priority changes, that could also change leadership. So we have to make sure that we make this clear in the documentation/tutorial so that the users know that.

Reply:
Another thing that is less trivial than it first appears with flexible quorums is what to do when we detect a higher ballot. If we detect a higher ballot that also `is_prepare_quorum()`, that's fine. But if we detect a higher ballot that, say, only `is_accept_quorum` but not `is_prepare_quorum`, it becomes less clear what to do. That server may or may not have completed the prepare phase. If it has completed the prepare phase, then it is correct to treat it as the leader. If not, then we will incorrectly continue to follow this stale leader.

I can think of two options on how to solve this, each with their own pros/cons:

1. Do an `is_prepare_quorum` check before we start following a higher ballot. This is the safest option but might cause unnecessary leader changes e.g., a leader was in accept phase and committing entries and then reconnects to another server that unnecessarily causes a leader change because the leader does not pass `is_prepare_quorum` check.
2. Incorporate more metadata in the ballots, such as if the servers are in the Prepare or Accept phase. This can probably avoid the unnecessary leader changes in (1) but starts to intermix BLE and Sequence Paxos logic.

At this point, I would prefer (1). The second option needs more investigation and proof of correctness before I feel comfortable to possibly merge it.